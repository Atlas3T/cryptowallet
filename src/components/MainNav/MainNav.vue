<template>
  <div class="columns is-mobile main-nav is-gapless">
    <div class="column">
      <router-link
        class="main-nav-link"
        to="/balance"
      >
        <q-icon name="fas fa-wallet" />
        <div>Balance</div>
      </router-link>
    </div>
    <div class="column">
      <router-link
        class="main-nav-link"
        to="/analytics"
      >
        <q-icon name="fas fa-chart-bar" />
        <div>Analytics</div>
      </router-link>
    </div>
    <div class="column">
      <router-link
        class="main-nav-link"
        to="/payments"
      >
        <q-icon name="fas fa-exchange-alt" />
        <div>Payments</div>
      </router-link>
    </div>
    <div class="column">
      <router-link
        class="main-nav-link"
        to="/exchange"
      >
        <q-icon name="fas fa-sync" />
        <div>Exchange</div>
      </router-link>
    </div>
    <div class="column">
      <router-link
        class="main-nav-link"
        to="/more"
      >
        <q-icon name="fas fa-ellipsis-h" />
        <div>More</div>
      </router-link>

    </div>
  </div>
</template>

<script>
export default {
  name: 'MainNav',
};
</script>

<style scoped>
  .main-nav {
    width: 100%;
    height: 15vh;
  }

  .main-nav > .column {
    display: flex;
    justify-content: center;
    align-items: center;
    padding: 5px;
  }

  .main-nav-link {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    color: black;
  }

  .main-nav-link i {
    margin-top: 0.3rem;
    font-size: 2.5rem;
  }

  .main-nav-link > div {
    margin-top: 0.5rem;
    font-size: 0.8rem;
  }
</style>
