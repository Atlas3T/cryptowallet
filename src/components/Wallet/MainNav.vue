<template>
  <section
    class="main-nav"
  >
    <div class="">
      <router-link
        class="main-nav-link"
        to="/wallet/wallet"
      >
        <i class="svg-icon">
          <img src="~assets/icons/Wallet.svg">
        </i>
        <div>{{ $t('wallet') }}</div>
      </router-link>
    </div>
    <div class="">
      <router-link
        class="main-nav-link"
        to="/wallet/analytics"
      >
        <i class="svg-icon">
          <img src="~assets/icons/Analytics.svg">
        </i>
        <div>{{ $t('analytics') }}</div>
      </router-link>
    </div>
    <div class="">
      <router-link
        class="main-nav-link"
        to="/wallet/payments"
      >
        <i class="svg-icon">
          <img src="~assets/icons/Payments.svg">
        </i>
        <div>{{ $t('payments') }}</div>
      </router-link>
    </div>
    <div class="">
      <router-link
        class="main-nav-link"
        to="/wallet/exchange"
      >
        <i class="svg-icon">
          <img src="~assets/icons/Exchange.svg">
        </i>
        <div>{{ $t('exchange') }}</div>
      </router-link>
    </div>
    <div class="">
      <router-link
        class="main-nav-link"
        to="/wallet/more"
      >
        <i class="svg-icon">
          <img src="~assets/icons/More.svg">
        </i>
        <div>{{ $t('more') }}</div>
      </router-link>

    </div>
  </section>
</template>

<script>
export default {
  name: 'MainNav',
  data() {
    return {
      icon: '/assets/icons/Analytics.svg',
    };
  },
};
</script>

<style scoped>
  .main-nav {
    width: 100%;
    height: 6rem;
    background: white;
    background: #f3f3f3;
    display: flex;
    justify-content: space-between;
  }

  .main-nav > div {
    flex: 1;
  }

  .main-nav-link {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    color: black;
    width: 100%;
    font-family: 'Montserrat-Regular';
    height: 100%;
  }

  .main-nav-link i {
    font-size: 2.5rem;
  }

  .main-nav-link > div {
    font-size: 0.8rem;
  }

  .router-link-active {
    background: #eaeaea;
  }

  .svg-icon {
    width: 100%;
    height: 4rem;
    line-height: 0;
    display: flex;
  }

  .svg-icon img {
    max-width: 75%;
    margin: 0 auto;
  }
</style>
