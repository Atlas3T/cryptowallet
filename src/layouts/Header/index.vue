<template>
  <section
    v-if="!hideHeader"
    class="header-section"
  >
    <div class="header-back-button-wrapper">
      <q-btn
        :disable="!isBackButtonEnabled"
        icon="arrow_back"
        color="primary"
        size="lg"
        class="icon-btn back-arrow-btn"
        flat
        @click.prevent="goBack"
      />
    </div>

    <div v-if="coinHeading">
      <img
        :src="coinLogo"
        class="coin-logo"
      >
      {{ wallet.displayName }}
    </div>
    <div v-else>
      <div v-if="heading === 'CryptoWallet'">
        <h1 class="header-h1 logo">{{ heading }}</h1>
      </div>
      <div v-else>
        <h1 class="header-h1">{{ heading }}</h1>
      </div>
    </div>

    <div
      v-if="displaySettings"
      class="header-settings-button-wrapper"
    >
      <q-btn
        :label="$t('settings')"
        color="secondary"
        size="sm"
        @click.prevent="goToSettings"
      />
    </div>

    <div
      v-if="displayAddWallet"
      class="header-settings-button-wrapper"
    >
      <q-btn
        icon="add"
        color="primary"
        size="lg"
        class="icon-btn icon-btn-right"
        flat
        @click.prevent="openWalletsModal"
      />
    </div>

    <div
      v-if="displayPriceChart"
      class="header-settings-button-wrapper"
    >
      <q-btn
        icon="timeline"
        color="primary"
        size="lg"
        class="icon-btn icon-btn-right"
        flat
        @click.prevent="openChartModal"
      />
    </div>

    <div
      v-if="displayAccounts"
      class="header-accounts-button-wrapper"
    >
      <q-btn
        icon="people"
        color="primary"
        size="lg"
        class="icon-btn icon-btn-right"
        flat
        @click.prevent="setAccountModalOpened(true)"
      />
    </div>


  </section>
</template>

<script>
import { mapState } from 'vuex';
import SelectAccount from '@/components/Modals/SelectAccount';

export default {
  name: 'Header',
  components: {
    SelectAccount,
  },
  data() {
    return {
      isBackButtonEnabled: false,
    };
  },
  computed: {
    ...mapState({
      id: state => state.route.params.id,
      authenticatedAccount: state => state.settings.authenticatedAccount,
    }),
    wallet() {
      return this.$store.getters['entities/wallet/find'](this.id);
    },
    displaySettings() {
      // if (this.authenticatedAccount) return true;
      return false;
    },
    displayAccounts() {
      return this.$route.path === '/';
    },
    displayAddWallet() {
      return this.$route.path === '/wallet';
    },
    displayPriceChart() {
      return this.$route.name === 'walletSingle';
    },
    supportedCoins() {
      return this.$store.state.settings.supportedCoins;
    },
    heading() {
      if (this.$route.name === 'setup') return '';
      if (this.$route.name === 'exchange') return 'Exchange';
      if (this.$route.name === 'settings') return 'Settings';
      return 'CryptoWallet';
    },
    coinHeading() {
      if (this.$route.name === 'walletSingle' ||
          this.$route.name === 'coinSinglePrices') {
        return true;
      }
      return false;
    },
    hideHeader() {
      if (this.$route.path === '/setup/0') return true;
      return false;
    },
    coinLogo() {
      const coin = this.supportedCoins.find(cc => cc.name === this.wallet.name);
      /* eslint-disable-next-line */
      return require(`@/assets/cc-icons/color/${coin.symbol.toLowerCase()}.svg`);
    },
  },
  watch: {
    $route() {
      if (window.history.length > 0 &&
          this.$route.path !== '/setup/0' &&
          this.$route.path !== '/' &&
          this.$route.path !== '/wallet') {
        this.isBackButtonEnabled = true;
      } else {
        this.isBackButtonEnabled = false;
      }
    },
  },
  methods: {
    goBack() {
      this.$router.go(-1);
    },
    goToSettings() {
      this.$router.push({ path: '/settings' });
    },
    setAccountModalOpened(value) {
      this.$root.$emit('selectAccountModalOpened', value);
    },
    openWalletsModal() {
      this.$root.$emit('walletsModalOpened', true);
    },
    openChartModal() {
      this.$router.push({ path: `/wallet/single/prices/${this.wallet.id}` });
    },
  },
};
</script>

<style>
.header-section {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 2.5rem;
  font-family: 'CooperHewitt-Semibold';
  background: #1e3c57;
  position: relative;
}

.header-h1 {
  letter-spacing: normal;
  font-size: 1em;
  margin-top: 0.3em;
  position: relative;
}

.header-h1.logo {
  font-family: 'CooperHewitt-BoldItalic';
}

.header-back-button-wrapper {
  position: absolute;
  left: 0.5rem;
  height: 100%;
  display: flex;
  align-items: center;
}

.header-settings-button-wrapper,
.header-accounts-button-wrapper {
  position: absolute;
  right: 0.5rem;
  height: 100%;
  display: flex;
  align-items: center;
}

.header-accounts-button-wrapper {
  right: .8rem;
}

.icon-btn {
  margin: 0;
  padding-bottom: 0;
  min-height: auto;
  height: 100%;
  position: relative;
  width: 4em;
  padding: 0;
}

.back-arrow-btn .q-btn-inner {
  justify-content: flex-start;
}

.icon-btn-right .q-btn-inner {
  justify-content: flex-end;
}

.back-arrow-btn .q-focus-helper {
  display: none;
}

.coin-logo {
  height: 1.25rem;
  margin-right: 0.2rem;
  position: relative;
  top: 0.2em;
}
</style>
