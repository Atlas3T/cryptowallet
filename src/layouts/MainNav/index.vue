<template>
  <section
    class="main-nav"
  >
    <div class="">
      <router-link
        class="main-nav-link"
        to="/wallet"
      >
        <q-icon name="account_balance_wallet" />
        <div>{{ $t('wallets') }}</div>
      </router-link>
    </div>
    <div class="">
      <router-link
        class="main-nav-link"
        to="/settings"
      >
        <q-icon name="settings" />
        <div>{{ $t('settings') }}</div>
      </router-link>
    </div>
  </section>
</template>

<script>
export default {
  name: 'MainNav',
};
</script>

<style scoped>
  .main-nav {
    width: 100%;
    height: 5rem;
    background: #0a2d3e;
    display: flex;
    justify-content: space-between;
    border-top: 0.3rem solid #4e677d;
    text-shadow: 1px 1px 0px rgba(0, 0, 0, 0.5);
  }

  .main-nav > div {
    flex: 1;
  }

  .main-nav-link {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    color: white;
    width: 100%;
    font-family: 'Montserrat-Medium';
    height: 100%;
    text-decoration: none;
  }

  .main-nav-link i {
    font-size: 2em;
    margin-bottom: 0.2em;
  }

  .main-nav-link > div {
    font-size: 0.8rem;
  }

  .router-link-exact-active {
    color: #16ac9f;
    font-family: 'Montserrat-Medium';
  }

  .svg-icon {
    width: 100%;
    height: 4rem;
    line-height: 0;
    display: flex;
  }

  .svg-icon img {
    max-width: 75%;
    margin: 0 auto;
  }
</style>
