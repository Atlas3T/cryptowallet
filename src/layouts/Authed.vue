<template>
  <q-layout view="lHh Lpr lFf">
    <q-page-container>
      <router-view/>
    </q-page-container>
    <transition
      enter-active-class="animated slideInUp"
      leave-active-class="animated slideOutDown"
    >
      <q-layout-footer v-show="isMainNavVisible">
        <MainNav/>
      </q-layout-footer>
    </transition>
  </q-layout>
</template>

<script>
import { mapState } from 'vuex';
import MainNav from '../components/Wallet/MainNav.vue';

export default {
  name: 'AuthedLayout',
  components: {
    MainNav,
  },
  data() {
    return {
      isMainNavVisible: false,
      transitionName: 'slide-left',
    };
  },
  computed: {
    ...mapState({
      isSearchingContacts: state => state.payments.isSearchingContacts,
    }),
  },
  watch: {
    $route(to, from) {
      this.updateMainNavVisiblity();
      const toDepth = to.path.split('/').length;
      const fromDepth = from.path.split('/').length;
      this.transitionName = toDepth < fromDepth ? 'slide-right' : 'slide-left';
    },
    isSearchingContacts() { this.updateMainNavVisiblity(); },
  },
  beforeMount() {
    this.updateMainNavVisiblity();
  },
  methods: {
    updateMainNavVisiblity() {
      if (this.$route.path === '/wallet/payments' && this.isSearchingContacts) {
        this.isMainNavVisible = false;
      } else {
        this.isMainNavVisible = true;
      }
    },
  },
};
</script>

<style>
.animated {
  animation-duration: 250ms;
}
</style>
