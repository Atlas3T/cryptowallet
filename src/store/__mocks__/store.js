import Vue from 'vue';
import Vuex from 'vuex';
import VuexORM from '@vuex-orm/core';
import VuexORMLoki from '@atlascity/vuex-orm-plugin-lokijs';

import wallet from '@/store/wallet';
import contacts from '@/store/contacts/__mocks__/contacts.js';
import search from '@/store/search/__mocks__/search.js';
import settings from '@/store/settings/__mocks__/settings.js';
import setup from '@/store/setup/__mocks__/setup.js';
import qrcode from '@/store/qrcode/__mocks__/qrcode.js';
import modals from '@/store/modals/__mocks__/modals.js';

// entities
import Account from '@/store/wallet/entities/account';
import Address from '@/store/wallet/entities/address';
import Tx from '@/store/wallet/entities/tx';
import Utxo from '@/store/wallet/entities/utxo';
import Wallet from '@/store/wallet/entities/wallet';
import Coin from '@/store/wallet/entities/coin';
import KeyPair from '@/store/wallet/entities/keyPair';
import Prices from '@/store/prices';
import LatestPrice from '@/store/latestPrice';

Vue.use(Vuex);

// Setup ORM database.
const database = new VuexORM.Database();

database.register(Account, {});
database.register(Address, {});
database.register(Tx, {});
database.register(Utxo, {});
database.register(Wallet, {});
database.register(Coin, {});
database.register(KeyPair, {});
database.register(Prices, {});
database.register(LatestPrice, {});

Vue.prototype.encryptedModels = [
  Account,
  Wallet,
];

const options = {
  env: 'browser',
};

VuexORM.use(VuexORMLoki, { database, options });

export const actions = {};
export const getters = {};
export const mutations = {};
export const state = {};

export function createMocks(custom = {
  getters: {}, mutations: {}, actions: {}, state: {},
}) {
  /**
   * mock store modules.
   */
  const modules = {
    wallet,
    search,
    settings,
    contacts,
    setup,
    qrcode,
    modals,
  };

  let mockGettersCombined = {};
  let mockMutationsCombined = {};
  let mockActionsCombined = {};
  let mockStateCombined = {};

  Object.entries(modules).forEach(([name, module]) => {
    mockGettersCombined = Object.assign({}, mockGettersCombined, module.getters);
    mockMutationsCombined = Object.assign({}, mockMutationsCombined, module.mutations);
    mockActionsCombined = Object.assign({}, mockActionsCombined, module.actions);
    mockStateCombined = Object.assign({}, mockStateCombined, module.state);

    if (custom.state && custom.state[name]) {
      module.state = Object.assign(module.state, custom.state[name]);
    }

    modules[name] = {
      namespaced: true,
      getters: Object.assign({}, getters, module.getters),
      mutations: Object.assign({}, mutations, module.mutations),
      actions: Object.assign({}, actions, module.actions),
      state: Object.assign({}, state, module.state),
    };
  });

  mockGettersCombined = Object.assign({}, mockGettersCombined, custom.getters);
  mockMutationsCombined = Object.assign({}, mockMutationsCombined, custom.mutations);
  mockActionsCombined = Object.assign({}, mockActionsCombined, custom.actions);
  mockStateCombined = Object.assign({}, mockStateCombined, custom.state);
  return {
    getters: mockGettersCombined,
    mutations: mockMutationsCombined,
    actions: mockActionsCombined,
    state: mockStateCombined,
    store: new Vuex.Store({
      getters: mockGettersCombined,
      mutations: mockMutationsCombined,
      actions: mockActionsCombined,
      state: mockStateCombined,
      modules,
      plugins: [VuexORM.install(database)],
    }),
  };
}

export const { store } = createMocks();
