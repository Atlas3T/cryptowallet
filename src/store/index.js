/*eslint-disable*/

import Vue from 'vue';
import Vuex from 'vuex';
import VuexPersistence from 'vuex-persist';
import VuexORM from '@vuex-orm/core';
import VuexORMLoki from 'vuex-orm-lokijs';

// import entities.
import Account from './wallet/entities/account';
import Address from './wallet/entities/address';
import Tx from './wallet/entities/tx';
import Utxo from './wallet/entities/utxo';
import Wallet from './wallet/entities/wallet';
import Coin from './wallet/entities/coin';
import KeyPair from './wallet/entities/keyPair';
import Prices from './prices';
import LatestPrice from './latestPrice';

// import modules.
import payments from './payments';
import search from './search';
import settings from './settings';
import setup from './setup';
import contacts from './contacts';

Vue.use(Vuex);

// Setup ORM database.
const database = new VuexORM.Database();

database.register(Account, {});
database.register(Address, {});
database.register(Tx, {});
database.register(Utxo, {});
database.register(Wallet, {});
database.register(Coin, {});
database.register(KeyPair, {});
database.register(Prices, {});
database.register(LatestPrice, {});

Vue.prototype.encryptedModels = [
  Account,
];

window.Tx = Tx;
window.Utxo = Utxo;
window.Address = Address;
window.Wallet = Wallet;
window.LatestPrice = LatestPrice;
window.Prices = Prices;
window.Account = Account;
window.Coin = Coin;

// Setup persistant storage.
const vuexLocal = new VuexPersistence({
  supportCircular: true,
  storage: window.localStorage,
  modules: [settings],
});


const options = {
  env: 'browser',
};

function hydrationCompletedCallback() {
  setTimeout(() => {
    store.dispatch('settings/setLoading', false);
  }, 1000);
}

VuexORM.use(VuexORMLoki, { database, options, hydrationCompletedCallback });


/**
 * Create CryptoWallet Vuex store obj.
 * @type {Store<any>}
 */
const store = new Vuex.Store({
  modules: {
    payments,
    search,
    settings,
    setup,
    contacts,
  },
  plugins: [VuexORM.install(database)],
});

Vue.prototype.activeWallets = {};

// @todo remove line below (used for development purposes only)
window.store = store;

export default store;
