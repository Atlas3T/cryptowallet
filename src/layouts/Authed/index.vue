<template>
  <q-layout
    :class="{
      'no-balance': !isBalanceVisible,
      'short-top': shortTop,
      'single-wallet-top': singleWalletTop,
    }"
    view="lHh Lpr lFf"
  >
    <Header />

    <div class="q-pull-to-refresh-wrapper">
      <q-pull-to-refresh
        :handler="refresher"
        :disable="!isPullEnabled || isPullTempDisabled"
        color="cyan"
      >
        <div class="background" />
        <div
          :class="{ 'no-footer': !isMainNavVisible }"
          class="layout-wrapper"
          @touchmove="prevent"
        >
          <div
            v-if="showCoinHeader"
            class="coin-header-wrapper"
          >
            <CoinHeader
              :wallet="wallet"
              :simple="true"
            />
          </div>

          <div class="total-balance-wrapper">
            <div
              v-if="showTotalBalance"
              class="total-balance"
            >
              <div>{{ totalBalance }}</div>
            </div>
          </div>

          <div
            :class="{ white: layoutShapeWhite }"
            class="layout-shape"
          >
            <keep-alive>
              <router-view />
            </keep-alive>
          </div>
        </div>
      </q-pull-to-refresh>
    </div>

    <transition
      enter-active-class="animated slideInUp"
      leave-active-class="animated slideOutDown"
    >
      <q-footer v-show="isMainNavVisible">
        <MainNav />
      </q-footer>
    </transition>
  </q-layout>
</template>

<script>
import { mapState } from 'vuex';
import Wallet from '@/store/wallet/entities/wallet';
import AmountFormatter from '@/helpers/AmountFormatter';
import MainNav from '@/layouts/MainNav';
import Header from '@/layouts/Header';
import CoinHeader from '@/components/Wallet/CoinHeader';
import Address from '@/store/wallet/entities/address';
import Utxo from '@/store/wallet/entities/utxo';

export default {
  name: 'AuthedLayout',
  components: {
    MainNav,
    Header,
    CoinHeader,
  },

  data() {
    return {
      isMainNavVisible: false,
      isPullEnabled: true,
      isPullTempDisabled: false,
      transitionName: 'slide-left',
      isBalanceVisible: true,
    };
  },

  computed: {
    ...mapState({
      id: (state) => { return state.route.params.id; },
      isSearchingContacts: (state) => { return state.search.isSearchingContacts; },
      authenticatedAccount: (state) => { return state.settings.authenticatedAccount; },
    }),

    wallets() {
      return Wallet.query()
        .where('account_id', this.authenticatedAccount)
        .where('imported', true)
        .get();
    },

    wallet() {
      return this.$store.getters['entities/wallet/find'](this.id);
    },

    selectedCurrency() {
      return this.$store.state.settings.selectedCurrency;
    },

    latestPrice() {
      const prices = this.$store.getters['entities/latestPrice/all'];
      return prices;
    },

    totalBalance() {
      let balance = 0;
      this.wallets.forEach((wallet) => {
        const price = this.$store.getters['entities/latestPrice/find'](`${wallet.symbol}_${this.selectedCurrency.code}`);
        if (price) {
          const formattedAmount = new AmountFormatter({
            amount: wallet.confirmedBalance,
            rate: price.data.PRICE,
            format: '0.00',
            coin: wallet.name,
            prependPlusOrMinus: false,
            currency: this.selectedCurrency,
            toCurrency: true,
            toCoin: false,
            withCurrencySymbol: false,
          });

          balance += parseFloat(formattedAmount.getFormatted());
        }
      });

      const formattedBalance = new AmountFormatter({
        amount: balance,
        format: '0,0[.]00',
        currency: this.selectedCurrency,
        toCurrency: false,
        toCoin: false,
        withCurrencySymbol: true,
      });

      return formattedBalance.getFormatted();
    },
    showTotalBalance() {
      return this.$route.name === 'wallet'
             || this.$route.name === 'sendCoin'
             || this.$route.name === 'receiveCoin'
             || this.$route.name === 'coinHistory'
             || this.$route.name === 'coinPrices';
    },
    layoutShapeWhite() {
      return this.$route.name === 'settings'
             || this.$route.name === 'exchange'
             || this.$route.name === 'walletSingle'
             || this.$route.name === 'sendCoinSingle'
             || this.$route.name === 'coinSinglePrices';
    },
    shortTop() {
      return this.$route.name === 'settings'
             || this.$route.name === 'exchange';
    },
    showCoinHeader() {
      return (this.$route.name === 'walletSingle' && this.wallet)
             || (this.$route.name === 'sendCoinSingle' && this.wallet)
             || (this.$route.name === 'coinSinglePrices' && this.wallet);
    },
    singleWalletTop() {
      return (this.$route.name === 'walletSingle' && this.wallet)
             || (this.$route.name === 'sendCoinSingle' && this.wallet)
             || (this.$route.name === 'coinSinglePrices' && this.wallet);
    },
  },

  watch: {
    /**
     * Watch the $route to keep the MainNav visibility updated
     */
    '$route.path': {
      handler() {
        this.updateMainNavVisibility();
        this.updateisPullEnabled();
      },
    },

    /**
     * Watch the isSearchingContacts to keep the MainNav visibility updated
     */
    isSearchingContacts() { this.updateMainNavVisibility(); },
  },

  /**
   * Make sure that the MainNav visibility is set correctly on initial render
   */
  beforeMount() {
    this.updateMainNavVisibility();
    this.updateisPullEnabled();
  },

  mounted() {
    this.$root.$on('isHomeBalanceVisible', (value) => {
      this.isBalanceVisible = value;
    });
  },

  methods: {
    prevent() {
      return false;
    },

    /**
     * Decide if the MainNav component should be visible
     * Update MainNav component visibility
     * @TODO James review future
     */
    updateMainNavVisibility() {
      if (this.$route.path === '/wallet/payments' && this.isSearchingContacts) {
        this.isMainNavVisible = false;
      } else {
        this.isMainNavVisible = true;
      }
    },

    updateisPullEnabled() {
      this.isPullEnabled = this.$route.name === 'wallet' || this.$route.name === 'walletSingle';
    },

    refresher(done) {
      if (this.$route.name === 'wallet') {
        setTimeout(async () => {
          try {
            await this.backEndService.loadPriceFeed();
            await this.updateBalances(done);
          } catch (err) {
            this.errorHandler(err);
            done();
          }
        }, 1000);
        return false;
      }

      if (this.$route.name === 'walletSingle') {
        setTimeout(() => {
          this.$root.$emit('updateWalletSingle', done);
        }, 1000);
      }

      return false;
    },

    async getUtxos(combinedAddresses, wallet) {
      if (wallet.sdk) {
        const coinSDK = this.coinSDKS[wallet.sdk];
        const utxos = await coinSDK.getUTXOs(combinedAddresses, wallet.network);

        let balance = 0;
        utxos.forEach((utxo) => {
          balance += utxo.amount;
          const found = Utxo.query()
            .where('txid', utxo.txid)
            .where('vout', utxo.vout)
            .where('wallet_id', wallet.id)
            .get();

          if (!found[0]) {
            utxo.account_id = this.authenticatedAccount;
            utxo.wallet_id = wallet.id;
            Utxo.$insert({ data: utxo });
          }
        });

        return {
          utxos,
          balance,
        };
      }
      return {};
    },

    async updateBalances(done) {
      const promises = [];

      function onlyUnique(value, index, self) {
        return self.indexOf(value) === index;
      }

      this.wallets.forEach((wallet) => {
        promises.push(new Promise(async (resolve, reject) => {
          try {
            const coinSDK = this.coinSDKS[wallet.sdk];

            const addresses = Address.query()
              .where('account_id', this.authenticatedAccount)
              .where('wallet_id', wallet.id)
              .where('used', false)
              .get();

            let addressesRaw = addresses.map((item) => { return item.address; });

            addressesRaw = addressesRaw.filter(onlyUnique);

            let newBalance;
            if (wallet.sdk === 'Bitcoin') {
              const result = await this.getUtxos(addressesRaw, wallet);
              const { balance } = result;
              newBalance = balance;
            } else if (wallet.sdk === 'Ethereum') {
              newBalance = await coinSDK.getBalance(addressesRaw, wallet.network);
            } else if (wallet.sdk === 'ERC20') {
              const activeWallet = this.activeWallets[this.authenticatedAccount][wallet.name];
              newBalance = await coinSDK.getBalance(activeWallet);
            }

            // update balance
            Wallet.$update({
              where: (record) => { return record.id === wallet.id; },
              data: { confirmedBalance: parseFloat(newBalance, 10) },
            });

            resolve();
          } catch (err) {
            reject(err);
          }
        }));
      });

      try {
        await Promise.all(promises);
        done();
      } catch (err) {
        this.errorHandler(err);
        done();
      }
    },
  },
};
</script>

<style>
.animated {
  animation-duration: 150ms;
}

.total-balance {
  display: flex;
  align-items: center;
  justify-content: center;
  color: white;
  font-family: 'CooperHewitt-BoldItalic';
  font-size: 2rem;
  line-height: 1;
  text-shadow: 1px 1px 0px rgba(0, 0, 0, 0.5);
  height: 5rem!important;
}

.total-balance > div {
  margin-top: 0.2em;
}

.total-balance-wrapper {
  transition: all ease-in-out 150ms;
  position: relative;
  top: 0;
  display: block!important;
  height: 5rem!important;
  position: absolute;
  width: 100%;
  opacity: 1;
  padding: 0 0.5rem;
}

.no-balance .total-balance-wrapper,
.short-top .total-balance-wrapper {
  top: -5rem;
  height: 0rem!important;
  opacity: 0;
}

.no-balance .scroll-area.extended {
  height: calc(100%);
}

.pull-to-refresh-message {
  transition: all ease-in-out 150ms;
  top: 0;
}

/*.no-balance .pull-to-refresh-message {
  top: -5rem!important;
}*/

.layout-shape .layout {
  height: calc(100vh - 5rem - 2.5rem);
}

.layout-shape.white {
  background: white!important;
}

.no-balance .layout-shape {
  background: #e4e9ef;
}

.background {
  transition: all ease-in-out 150ms;
}

.no-balance .background,
.short-top .background {
  height: 21rem;
}

.single-wallet-top .background {
  height: 28.5rem;
}

.q-pull-to-refresh-wrapper {
  padding-top: 2.5rem;
}

.q-pull-to-refresh-wrapper .pull-to-refresh-message {
  position: relative;
  z-index: 2
}

.coin-header-wrapper {
  height: 8rem!important;
  padding: 1rem 0.5rem;
}

.single-wallet-top .coin-header-wrapper {
  height: 10rem!important;
  display: flex;
  align-items: center;
  justify-content: center;
  padding-bottom: 0rem;
}

</style>
