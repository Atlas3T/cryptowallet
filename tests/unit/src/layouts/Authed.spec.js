import { shallowMount } from '@vue/test-utils';
import Authed from '@/layouts/Authed.vue';
import Header from '@/components/Wallet/Header.vue';
import MainNav from '@/components/Wallet/MainNav.vue'
import { localVue, i18n, createRouter } from '@/../tests/unit/helpers/setupLocalVue.js';
import { __createMocks as createStoreMocks } from '@/../tests/store/__mocks__/store.js';

describe('Authed.vue', () => {
  let storeMocks;
  let wrapper;
  let store;
  let router;

  function wrapperInit (options) {
    return shallowMount(Authed, options);
  }

  function storeInit (custom) {
    storeMocks = createStoreMocks(custom);
    router = createRouter(storeMocks.store);
    router.push({ path: `/` });
    wrapper = wrapperInit({ i18n, router, localVue, store: storeMocks.store });
    store = wrapper.vm.$store;
  }

  beforeEach(() => storeInit());

  it('renders and matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders a Header component', () => {
    expect(wrapper.contains(Header)).toBe(true);
  });

  it('renders a router-view component', () => {
    expect(wrapper.contains('router-view-stub')).toBe(true);
  });

  it('renders a MainNav component', () => {
    expect(wrapper.contains(MainNav)).toBe(true);
  });

  it('calls updateMainNavVisibility() on route change', () => {
    const updateMainNavVisibilityMock = jest.fn();
    wrapper.setMethods({updateMainNavVisibility: updateMainNavVisibilityMock});
    router.push({ path: `/fake1` });
    router.push({ path: `/fake2` });
    expect(updateMainNavVisibilityMock).toHaveBeenCalledTimes(2);
  });

  it('calls updateMainNavVisibility() on isSearchingContacts change', () => {
    const updateMainNavVisibilityMock = jest.fn();
    wrapper.setMethods({updateMainNavVisibility: updateMainNavVisibilityMock});
    store.state.search.isSearchingContacts = true;
    expect(updateMainNavVisibilityMock).toHaveBeenCalledTimes(1);
    store.state.search.isSearchingContacts = false;
    expect(updateMainNavVisibilityMock).toHaveBeenCalledTimes(2);
  });

  it('hides MainNav if searching contacts and route is "/wallet/payments"', () => {
    expect(wrapper.find('qlayoutfooter-stub').isVisible()).toBe(true);
    store.state.search.isSearchingContacts = true;
    router.push({ path: `/wallet/payments` });
    expect(wrapper.find('qlayoutfooter-stub').isVisible()).toBe(false);
  });
});
