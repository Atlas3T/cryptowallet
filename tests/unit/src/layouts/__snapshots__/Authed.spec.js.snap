// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authed.vue renders and matches snapshot 1`] = `
<qlayout-stub
  view="lHh Lpr lFf"
>
  <qlayoutheader-stub
    revealoffset="250"
    value="true"
  >
    <header-stub />
  </qlayoutheader-stub>
   
  <div
    class="layout-wrapper"
  >
    <router-view-stub
      name="default"
    />
  </div>
   
  <qlayoutfooter-stub
    enter-active-class="animated slideInUp"
    leave-active-class="animated slideOutDown"
    value="true"
  >
    <mainnav-stub />
  </qlayoutfooter-stub>
</qlayout-stub>
`;
